app/models/solidus_stripe/payment_method.rb

```diff
@@ -7,37 +7,143 @@ class PaymentMethod < ::Spree::PaymentMethod
 
-    def partial_name
-      "stripe"
-    end
+    concerning :Actions do
+      def actions
+        %w[capture void credit]
+      end
 
-    alias cart_partial_name partial_name
-    alias product_page_partial_name partial_name
+      def can_capture?(payment)
+        payment.pending?
+      end
+
+      def can_void?(payment)
+        payment.pending?
+      end
 
-    def payment_source_class
-      PaymentSource
+      def can_credit?(payment)
+        payment.completed? && payment.credit_allowed > 0
+      end
     end
 
-    def gateway_class
-      Gateway
+    concerning :Configuration do
+      def partial_name
+        "stripe"
+      end
+
+      alias cart_partial_name partial_name
+      alias product_page_partial_name partial_name
+      alias risky_partial_name partial_name
+
+      def source_required?
+        false
+      end
+
+      def payment_source_class
+        PaymentSource
+      end
+
+      def gateway_class
+        Gateway
+      end
     end
 
-    def create_profile(payment)
-      payment_intent = payment.source.payment_intent
+    concerning :Order do
+      def find_in_progress_payment_for(order)
+        payments = order.payments
+          .where.not(state: %w[completed invalid void]) # in_progress
+          .where(payment_method: self)
+          .order(:created_at)
+          .entries
+
+        *old_payments, payment = payments
+        old_payments.each(&:invalidate!)
+
+        if payment && payment.amount != order.total
+          payment.cancel!
+          payment = nil
+        end
+
+        payment
+      end
+
+      def create_in_progress_payment_for(order)
+        transaction do
+          intent = gateway.request do
+            Stripe::PaymentIntent.create({
+              amount: gateway.to_stripe_amount(
+                order.display_total.money.fractional,
+                order.currency,
+              ),
+              currency: order.currency,
 
-      if payment_intent && payment_intent.customer.blank?
-        payment.payment_method.gateway.request do
-          payment_intent.customer = Stripe::Customer.new email: payment.order.email
-          payment_intent.customer.save
-          payment_intent.save
+              # The capture method should stay manual in order to
+              # avoid capturing the money before the order is completed.
+              capture_method: 'manual',
+            })
+          end
+
+          order.payments
+            .create!(
+              payment_method: self,
+              response_code: intent.id,
+              amount: order.total,
+            ).tap(&:started_processing!)
         end
       end
 
-      self
+      def find_or_create_in_progress_payment_for(order)
+        payment = find_in_progress_payment_for(order)
+        intent = find_intent_for(payment) if payment
+
+        payment = nil if intent.nil?
+        payment = nil unless intent&.status == 'requires_payment_method'
```

Comment:
From my tests here it seems like the intent you create during `create_in_progress_payment_for(order)` has always the status 'requires_source' (waiting to enter card details, I understand). So if you run multiple times `find_or_create_in_progress_payment_for` it will always recreate a payment + payment intent because you nulify the payment on this line. I think that's not what we want, we should void recreating intents whenever possible. What is the purpose of this line exactly ?

Reply:
Tracked in #201 thanks!